### ISSUE TITLE
WPF DataGrid not adding empty row in version 1.4.0

### ISSUE THREAD

**Issue by rubend30** [MintPlayer/MintPlayer.Dotnet.Tools#6]
Hello

In the current version (v1.4.0) it is not possible to add new rows in a WPF DataGrid. Now only one row can be added but no new empty row is available. The GIF below illustrates this issue.

I made an example project so you can test this issue:
[IssueObservableCollection20210713.zip](https://github.com/MintPlayer/MintPlayer.ObservableCollection/files/6807497/IssueObservableCollection20210713.zip)

This GIF also shows the problem. The left grid is the build-in ObservableCollection while the right is this Nuget package with version 1.4.0. I checked other versions and it only happens when in version 1.4.0.
[GIF of the issue](https://i.imgur.com/02ZdPxb.gif)

If I can I will try to find and fix this issue.

Thank you for any help and insights in this issue.

**Comment by rubend30** [comment]
What I don't really understand is the purpose of this code:

```
var isRange = IsRange(e);
if (isRange && isCollectionView)
{
    // Call the Refresh method if the target is a WPF CollectionView
    RunOnMainThread(
        (param) => handler.Target.GetType().GetMethod("Refresh").Invoke(param.target, new object[0]),
        new { target = handler.Target }
    );
}
else if (Enabled)
{
    RunOnMainThread(
        (param) => handler(this, param.e),
        new { e }
    );
}
```

It seems that the behaviour is correct for my issue when the code enters the `else if`-block but not when it enters the `if `-block.

And the method:

```
private bool IsRange(NotifyCollectionChangedEventArgs e)
{
    return e.NewItems?.Count != 1 || e.OldItems?.Count != 1;
}
```

What is exacly the idea behind this? When the `NewItems `or `OldItems` contain 0 items it is also a range? Why is having only one item not a range?

**Comment by PieterjanDeClippel** [comment]
By default, you cannot use `AddRange` in WPF. For the built-in ObservableCollection you'd get the error `Range operations are not supported`. This package fixes this by calling `InsertRange` on the underlying `List<T>`

https://github.com/MintPlayer/MintPlayer.ObservableCollection/blob/34504a69fae999dc2ec23cce8fbccde5038f3173/MintPlayer.ObservableCollection/ObservableCollection.cs#L271

 and iterating all event handlers associated with the CollectionChanged event

https://github.com/MintPlayer/MintPlayer.ObservableCollection/blob/34504a69fae999dc2ec23cce8fbccde5038f3173/MintPlayer.ObservableCollection/ObservableCollection.cs#L119

If the event handler's target is a `CollectionView`, you cannot invoke the handler, but instead it `Refresh`es the `CollectionView`

https://github.com/MintPlayer/MintPlayer.ObservableCollection/blob/34504a69fae999dc2ec23cce8fbccde5038f3173/MintPlayer.ObservableCollection/ObservableCollection.cs#L125-L132

> When the NewItems or OldItems contain 0 items it is also a range?

Yes, indeed. WPF would also throw an exception for a normal `ObservableCollection` when adding 0 items.

**Comment by PieterjanDeClippel** [comment]
Blijkbaar is bij het uitvoeren van de fout-reproductie, op volgende lijn in de library

https://github.com/MintPlayer/MintPlayer.ObservableCollection/blob/34504a69fae999dc2ec23cce8fbccde5038f3173/MintPlayer.ObservableCollection/ObservableCollection.cs#L305

    CollectionChanged == null

Maar in [volgende repository met voorbeeld](https://github.com/PieterjanDeClippel/WpfRangeOperations) heeft het `CollectionChanged` event wel gekoppelde handlers.

![image](https://user-images.githubusercontent.com/9629574/126638732-000c4cbb-0db6-4b01-b416-abb6018efd1b.png)

**Comment by PieterjanDeClippel** [comment]
Ik vermoed dat `Refresh` hier de oorzaak van is.

https://github.com/MintPlayer/MintPlayer.ObservableCollection/blob/34504a69fae999dc2ec23cce8fbccde5038f3173/MintPlayer.ObservableCollection/ObservableCollection.cs#L129

Vanaf het moment dat je begint met te typen in de laatste rij, wordt een item toegevoegd, en dus het `OnCollectionChanged` opgeroepen.

**Comment by PieterjanDeClippel** [comment]
I've released a new version 1.5.0. You can simply update the nuget package to version 1.5.0 to resolve the issue.

`IsRange` is now checked first. If there's only one item added/removed, I'm now just calling the `base.OnCollectionChanged` method.

Can you confirm if this resolves your problem? Dan kan dit issue gesloten worden...
